File: code/language/src/main/java/org/fbme/ide/iec61499/editor/ProjectEditorSpecs.java
package org.fbme.ide.iec61499.editor;

import jetbrains.mps.project.Project;
import org.jetbrains.mps.openapi.model.SNode;

import java.util.HashMap;
import java.util.Map;

public abstract class ProjectEditorSpecs {

    static final Map<Project, ProjectEditorSpecs> ourInstances = new HashMap<>();

    public static ProjectEditorSpecs getInstance(Project project) {
        return ourInstances.get(project);
    }

    public abstract SNode getSpec(SNode target, String projectionState);

    public abstract SNode getSpecTarget(SNode spec);

    public abstract String getSpecProjectionState(SNode spec);
}

File: code/language/src/main/kotlin/org/fbme/ide/iec61499/editor/ProjectEditorSpecs.kt
package org.fbme.ide.iec61499.editor

import jetbrains.mps.project.Project
import org.jetbrains.mps.openapi.model.SNode

abstract class ProjectEditorSpecs {
    abstract fun getSpec(target: SNode, projectionState: String): SNode
    abstract fun getSpecTarget(spec: SNode): SNode
    abstract fun getSpecProjectionState(spec: SNode): String?

    companion object {
        @JvmField
        val ourInstances: Map<Project, ProjectEditorSpecs> = HashMap()
        fun getInstance(project: Project): ProjectEditorSpecs {
            return ourInstances[project] ?: error("ProjectEditorSpecs not found")
        }
    }
}
